book/src/developers/designs/application_passwords.md

```diff
@@ -84,19 +83,18 @@ the UNIX one.
 
 ### Application Entries
 
-A new class for applications will be added that has an associated reference to a
-set of groups that are allowed access.
+A new class for applications will be added. Each application will have a single
+associated group so only members of this group will be able to bind with the
+application password for the associated application.
 
-(NOTE: It could be good to make this a group such that members of the group
-inherit "memberof=application X". This would make presentation of application
-password links in the UI easier since we can then do a search on (memberof &&
-class = application). Alternately we don't need this since we can in-memory
-cache applications and just compare their member sets to the user sessions
-memberof)
+(NOTE: Should we also disallow the creation of application passwords for an
+application if the user is not a member of the associated group? This will
+complicate reference integrity, as removing a member will require to remove the
+application password.)
```

Comment:
Referential integrity is to "clean up" references as a cascade when a delete occurs. It's NOT there to impose constraints on what is or isn't allowed to exist.

We can implement that logic in the idm layer so that on the api call to create a new application password can check if the user is a memberof the needed access groups. That way we can enforce this constraint at application password create time. 
The question *then* is how we handle the application password if the user is removed from the group that can access the application. IMO it should NOT trigger the delete. You can imagine a scenario where someone says "I lost my laptop, please disable my access temporarily". You can move them out the group and back in when they find the device for example. There are plenty of reasons to move the user out then into the application group if needed that shouldn't delete the application password. 

(It's also an argument that we should have booleans on each app password that says if that password is enabled, so that the user can selectively enable/disable application passwords).

Reply:
The "disable this password" thing feels like a too-many-knobs situation (and a "why not just delete and create a new one). 

- there's prior art in that individual AWS credentials can be disabled and re-enabled
- it'd be useful to test disabling a credential from a "scream testing" perspective, to see if it'll break anything in complex legacy workflows (which is why they're using LDAP in the first place).

On the delete thing, that should be a system-level cleanup task TBH, disable the credential when they're removed from the group and have an updated/last used flag on the cred, with a method to remove "floating" creds after x time.

Reply:
yep i agree. It's also easier to add later than remove something we don't need later.

Reply:
Ok. Please check the doc update to verify I have captured your ideas correctly.